### ISSUE TITLE
twisted.internet.process should not be importable on Windows

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6894]
|<img alt="koukopoulos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| koukopoulos reported|
|-|-|
|Trac ID|trac#6894|
|Type|defect|
|Created|2014-01-08 17:46:02Z|
|Branch|https://github.com/twisted/twisted/tree/no-windows-tiprocess-6894|

In _twisted/internet/process.py_ the _detectLinuxBrokenPipeBehavior_ is called unconditionally and fails on Windows systems:

```
...
  File "c:\users\konstantinos\documents\github\twisted\twisted\internet\process.py", line 103, in <module>
    detectLinuxBrokenPipeBehavior()
  File "c:\users\konstantinos\documents\github\twisted\twisted\internet\process.py", line 93, in detectLinuxBrokenPipeBehavior
    reads, writes, exes = select.select([w], [], [], 0)
select.error: (10038, 'An operation was attempted on something that is not a socket')
```

Attachments:

* [detect-broken-linux.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/16b/16b0f0d55b0c0057a6543e8d97f08fd3b85fd487/4cc0825ae387a9bfb3ff57362a3ebfe0cc2b7907.patch) (868 bytes) - added by koukopoulos on 2014-01-08 17:46:34Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6894 6894
type__defect defect
reporter__koukopoulos koukopoulos
priority__normal normal
milestone__ 
branch__branches_no_windows_tiprocess_6894 branches/no-windows-tiprocess-6894
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1389203162000000 1389203162000000
changetime__1444279803066407 1444279803066407
version__None None
owner__hawkowl hawkowl
cc__koukopoulos
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Hi koukopoulos.  Thanks for your contribution.

One thing to note is that `twisted/internet/process.py` isn't for use on Windows so I'm not sure this is actually a problem.

What is this change supposed to enable?

**Comment by twisted-trac** [comment]
|<img alt="koukopoulos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|koukopoulos commented|
|-|-|

You are right, there is nothing in twisted/internet/process.py usefull for Windows. However as I find is the case, a third-party library, Crochet, is importing the module. Apparently it is done as a workaround for some other issue with Twisted. Here are some relevant links:

https://github.com/itamarst/crochet/commit/780d33146fe5351bce15ac4e738d423dbe274869
http://twistedmatrix.com/trac/ticket/6738
http://twistedmatrix.com/trac/ticket/791
http://twistedmatrix.com/trac/ticket/733

It seems Crochet only calls methods of twisted.internet.process when running on POSIX. However the import happens unconditionally and that leads to detectLinuxBrokenPipeBehavior being called at import time even on Windows. I will ask the developer of Crochet to not import the module if on a non-posix platform but to me it seems that an easier and more logical fix is making it so that simply importing the module does not cause the program to crash.

**Comment by twisted-trac** [comment]
|<img alt="koukopoulos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|koukopoulos commented|
|-|-|

I opened an issue with Crochet as well: https://github.com/itamarst/crochet/issues/21

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

I think Crochet has only itself to blame for this. :)  (Notice the *Maintainer* documented for `process.py` and the author of Crochet)

It's true the attached patch is quite short and simple.  However, there's absolutely no testing of `twisted.internet.process` on Windows.  That means there's no guarantee it will continue to be importable after future changes.  If you wanted this reliably fixed, then you'd need to also write a unit test demonstrating `twisted.internet.process` can be imported without error on Windows.

At that point, it may no longer be simpler than fixing Crochet to not import this module on Windows.

**Comment by twisted-trac** [comment]
|<img alt="jjmontes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|jjmontes commented|
|-|-|

Hello.

I'm seeing the same issue, this time as the process module is imported by Autobahn.


```
  File "C:\Devel\console-web\env275w64\lib\site-packages\autobahn\twisted\wamp.py", line 40, in <module>
    from autobahn.twisted.websocket import WampWebSocketClientFactory
  File "C:\Devel\console-web\env275w64\lib\site-packages\autobahn\twisted\websocket.py", line 42, in <module>
    from autobahn.twisted.util import peer2str
  File "C:\Devel\console-web\env275w64\lib\site-packages\autobahn\twisted\util.py", line 31, in <module>
    from twisted.internet._posixstdio import PipeAddress
  File "C:\Devel\console-web\env275w64\lib\site-packages\twisted\internet\_posixstdio.py", line 16, in <module>
    from twisted.internet import process, error, interfaces
  File "C:\Devel\console-web\env275w64\lib\site-packages\twisted\internet\process.py", line 119, in <module>
    brokenLinuxPipeBehavior = detectLinuxBrokenPipeBehavior()
  File "C:\Devel\console-web\env275w64\lib\site-packages\twisted\internet\process.py", line 107, in detectLinuxBrokenPipeBehavior
    reads, writes, exes = select.select([w], [], [], 0)
select.error: (10038, 'An operation was attempted on something that is not a socket')

```

The same code works fine on Linux.

I have filed a report to Autobahn (https://github.com/tavendo/AutobahnPython/issues/499). However, I still feel that Twisted should avoid this _import_ failure when running under Windows.

The same application works if I patch process.py:

```
if sys.platform.startswith('linux'):
    brokenLinuxPipeBehavior = detectLinuxBrokenPipeBehavior()
else:
    brokenLinuxPipeBehavior = False
```

But obviously I'd prefer to avoid patching any of my dependencies :).

Best regards!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

This is an interesting issue -- ```twisted.internet.process``` doesn't import on Windows, but nobody ever should because it doesn't work on Windows. ```twisted.internet._dumbwin32proc``` is the API that is used when ```spawnProcess``` is called, not ```t.i.process```.

A better solution for this would be to just raise ```ImportError``` if you try and load it on Windows, because none of it will work on it.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

(In [45679]) Branching to no-windows-tiprocess-6894.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl commented|
|-|-|

Rather than provide a perplexing error message, this branch gives a helpful one.

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes look (almost) good :)

There is a pyflakes error:

```
twisted/internet/test/test_process.py:887: 'twisted' imported but unused
```

Please fix and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45725]) Merge no-windows-tiprocess-6894: t.internet.process shouldn't be importable on win32

Author: hawkowl
Reviewer: adiroiban
Fixes: [#6894](https://github.com/twisted/twisted/issues/6894)
